Fix Custom Fields Handling in Storage for both Orders and Products

pmusolino opened the pull request:
Closes #14106

## Description
It has been reported that for some users orders does not load peaMlT-Wz. It seems that this can happen when a store has a lot of custom fields for products, so upserting custom fields creates a deadlock on the writer context, delaying the syncing of orders for around 2 minutes.
This update optimizes the handling of custom fields in both `OrdersUpsertUseCase` and `ProductStore`. The changes improve performance by streamlining the logic for managing custom fields, introducing batch processing, and eliminating redundant operations.

## Summary of changes
- **OrdersUpsertUseCase**
  - Removed checks for the existence of custom fields in `storageOrder` that are not present in `readOnlyOrder`, directly removing them to simplify the logic.
  - Converted the creation of new custom fields from a loop with appends to a map for efficiency.
  - Implemented batch writing for new custom fields to reduce database calls.

- **ProductStore**
  - Improved readability and indentation for better code maintenance.
  - Simplified the logic for removing custom fields that are not in `readOnlyProduct` by directly using a set lookup.
  - Removed redundant operations for managing custom fields.
  - Introduced batch processing for adding new custom fields, thereby optimizing database interactions.

## Steps to reproduce
1. Login to a store with a lot of products and orders, with a lot of custom fields for each one (more than 100).
2. Notice that Order List is not loading, if not after a lot of time.
3. Make sure that in the product detail, the list of custom fields will be loaded correctly (cc @hafizrahman).

## Testing information
Tested that in a store with a lot of custom fields for products and orders, and everything loads properly.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

pmusolino committed:
- Add a set of metadata IDs for quick lookup in `handleProductCustomFields`
- Update comment to clarify the removal process of custom fields in `handleProductCustomFields`

pmusolino committed:
- Adds batch writing process for multiple custom fields in `ProductStore.swift`
- Creates a new array to store new metadata objects before adding to `storageProduct`

pmusolino committed:
- Remove redundant removal of storage custom fields not present in read-only product
- Update logic to only handle new custom fields from read-only product

pmusolino committed:
- Improve indentation and readability in ProductStore.swift
- Streamline logic for removing custom fields that are not in read-only product

pmusolino committed:
Refactor Order Custom Fields Handling in `OrdersUpsertUseCase`
- Optimize custom fields handling in OrdersUpsertUseCase.swift
- Replace forEach with a set lookup to improve performance
- Introduce batch writing for new custom fields to reduce database calls

itsmeichigo committed:
Remove all stored custom fields for

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  1 Message: This PR contains changes to <c...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/225">20.6 ❄️</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR contains changes to <code>RELEASE-NOTES.txt</code>.<br />
Note that these changes won't affect the final version of the release notes as this version is in code freeze.<br />
Please, get in touch with a release manager if you want to update the final release notes.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino committed:
update: release-notes 20.6

itsmeichigo commented on Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift:
```diff
@@ -304,25 +304,25 @@ struct OrdersUpsertUseCase {
+        let readOnlyMetadataIDs = Set(readOnlyOrder.customFields.map { $0.metadataID })
 
-        // Now, remove any objects that exist in `storageOrder.customFields` but not in `readOnlyOrder.customFields`
+        // Remove any objects that exist in `storageOrder.customFields` but not in `readOnlyOrder.customFields`
```
I'll let someone else review this as I also contributed a bit to it, but this comment (and the one for products) should be updated. We are removing all custom fields for the order right?

pmusolino replied:
Fixed here 9d34b5c8b8214eb7147ba806b809da1222700c96

pmusolino committed:
fix: comments

selanthiraiyan commented on Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift:
```diff
@@ -304,25 +304,25 @@ struct OrdersUpsertUseCase {
-            }
-        }
+        // Create a set of metadata IDs from the read-only order for quick lookup
+        let readOnlyMetadataIDs = Set(readOnlyOrder.customFields.map { $0.metadataID })
```
I think this isn't used anywhere.

pmusolino replied:
Fixed here d51ee596fb1c3c0533cacb338e2eb755c3c3bb8e 👍

selanthiraiyan commented on Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift:
```diff
@@ -304,25 +304,25 @@ struct OrdersUpsertUseCase {
+            return newStorageMetaData
+        }
+
+        // Batch writing process of multiple custom fields
```
Nice. 👍

selanthiraiyan commented on Yosemite/Yosemite/Stores/ProductStore.swift:
```diff
@@ -1162,25 +1162,25 @@ extension ProductStore {
-                storageProduct.addToCustomFields(newStorageMetaData)
-            }
-        }
+        // Create a set of metadata IDs from the read-only product for quick lookup
+        let readOnlyMetadataIDs = Set(readOnlyProduct.customFields.map { $0.metadataID })
```
I think this isn't used anywhere.

pmusolino replied:
d51ee596fb1c3c0533cacb338e2eb755c3c3bb8e

selanthiraiyan commented on Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift:
```diff
@@ -304,25 +304,25 @@ struct OrdersUpsertUseCase {
+            storage.deleteObject(storageCustomField)
+        }
+
+        // Create `customFields` objects from the `readOnlyOrder`
```
I think we should unit tests to `OrdersUpsertUseCaseTests` to validate that it can handle a large number of `MetaData` items. Right now, I see only tests that add/remove 1 custom field. What do you think?

pmusolino replied:
Implemented here dd9ca2356ceb1927721114c57c770a18409939ed

selanthiraiyan commented on Yosemite/Yosemite/Stores/ProductStore.swift:
```diff
@@ -1162,25 +1162,25 @@ extension ProductStore {
+            storage.deleteObject(storageCustomField)
+        }
+
+        // Create `customFields` objects from the `readOnlyProduct`
```
Similar to the other comment. I think we should unit tests to validate that it can handle a large number of MetaData items.

pmusolino replied:
The test implemented here dd9ca2356ceb1927721114c57c770a18409939ed should work also for Products

selanthiraiyan reviewed (commented):
Thanks for working on this, Paolo!

I tested this on a store with a large number of custom fields and I validated that orders are loaded as expected. ✅ 

I will wait for @hafizrahman to test that custom fields are displayed without issues.

hafizrahman commented:
> Make sure that in the product detail, the list of custom fields will be loaded correctly (cc @hafizrahman).

At this point in this branch, the Product Details is not able to actually show custom fields yet (only shows entry point), with the logic to show them is in a different branch in #14068

However, I was able to test this by putting a breakpoint inside `ProductFormViewController`. For each time I'm entering a Product Detail, I can confirm that `p viewModel.productModel` shows the content of the product and it includes its `customFields` property properly. Since it is what's used to display the content, I'm fairly confident there's no issue here.

Update: additionally, loading the product list screen itself also works fine. This is tested in both iPad and iPhone in iOS 17.5 Simulator with a site with many orders and products, where the products have many custom fields.

pmusolino committed:
cleanup unused code

selanthiraiyan commented:
@pmusolino Please update the PR description to close the issue #14106

pmusolino committed:
Add test for handling large number of custom fields in orders and products
- Modify `OrdersUpsertUseCaseTests.swift` to include a new test case for processing 5000 custom fields without deadlock.

pmusolino committed:
update: removed extra space

pmusolino commented:
@selanthiraiyan ready for another check

jaclync commented on Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift:
```diff
@@ -304,25 +304,23 @@ struct OrdersUpsertUseCase {
+
+        // Batch writing process of multiple custom fields
+        if !newStorageMetaDataArray.isEmpty {
+            storageOrder.addToCustomFields(NSSet(array: newStorageMetaDataArray))
```
@pmusolino Just sharing an observation from some debugging - this line caused the upsert to be triggered twice, with `newStorageMetaDataArray` containing a different custom field object (different Core Data object ID) each time. It's the expected result of the updated approach in this PR, with pre-existing custom fields being deleted and new custom fields re-added.

selanthiraiyan replied:
Thanks, Jaclyn!

From the comment of [the failing unit test](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Yosemite/YosemiteTests/Stores/OrderStoreTests.swift#L971) I wonder whether the updated approach of deleting all the stored custom fields before adding new custom fields will cause any reload or UI issues in screens where we use `EntityListener` to listen to changes to product or order entities. 🤔

selanthiraiyan reviewed (commented):
Thanks a lot for working on this @pmusolino and @itsmeichigo! I appreciate the efforts. 🙇 

With the new approach of deleting all the `MetaData` items from Storage we are not doing the "upsert" operation of orders and products. i.e. We are not following the "upsert" behaviour, which is to check the existing items from core data and update them. Instead, we delete all the `MetaData` items and add them again regardless of whether something changed. 

I think we should look for a different solution for the following reasons. 


1. [The existing method names and comments](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Yosemite/Yosemite/Stores/Order/OrdersUpsertUseCase.swift#L19-L25) expect an upsert operation instead of deleting everything. This is misleading. 
2. Core data will consider that the entity is changed even if we write the same order or product into the storage again. (which triggers callbacks and UI reloads) This is not expected from an "upsert" process. 

Unfortunately, I don't have any suggestions currently.

pmusolino committed:
Refactor custom fields handling in Orders and Products
- Update `OrdersUpsertUseCase.swift` to upsert custom fields from `readOnlyOrder`.
- Update `ProductStore.swift` to upsert custom fields from `readOnlyProduct`.
- Remove redundant custom fields from both `storageOrder` and `storageProduct` based on `readOnlyOrder` and `readOnlyProduct`.

pmusolino commented:
@selanthiraiyan I have pushed new changes here: 3cc83fdf4435a910761b0ebae87ec27c8c667f36. In this update, `MetaData` items will not be deleted; instead, there will be just the usual upsert. As a result, the previous unit test in `OrderStoreTests` will no longer fail. These new changes offer blazing fast performance compared to the previous approach while obtaining the same result. It's just an improvement on how metadata are handled in the main cycle. With these changes, I am no longer able to replicate the issue, and the unit test implemented in this PR still works as expected.
Please take a second look 🙌

selanthiraiyan commented on Yosemite/YosemiteTests/Stores/Order/OrdersUpsertUseCaseTests.swift:
```diff
@@ -262,6 +261,30 @@ final class OrdersUpsertUseCaseTests: XCTestCase {
+        DispatchQueue.global(qos: .background).async {
+            orderUseCase.upsert([order])
+            productStore.upsertStoredProducts(readOnlyProducts: [product], in: backgroundContext)
+        backgroundContext.saveIfNeeded()
```
nit: Indentation update needed.

pmusolino replied:
161f7cd5ba6decc8c5a91f26a3b00ac5f78cffd5

selanthiraiyan reviewed (approved):
This looks great! 👏 

I tested it using the store with issues, and the orders load as expected.

pmusolino committed:
fix: indentation
